fw/mmap_buffer.c

```diff
@@ -0,0 +1,329 @@
+}
+
+TfwMmapBufferHolder *
+tfw_mmap_buffer_create(const char *filename, unsigned int size)
+{
+	TfwMmapBufferHolder *holder;
+	struct page **page_ptr = NULL;
+	unsigned int order, i, page_cnt;
+	int cpu;
+
+	if (size < TFW_MMAP_BUFFER_MIN_SIZE
+	    || size > TFW_MMAP_BUFFER_MAX_SIZE
+	    || !is_power_of_2(size))
+		return NULL;
+
+	if (filename && strlen(filename) >= TFW_MMAP_BUFFER_MAX_NAME_LEN - 1)
+		return NULL;
+
+	if (holders_cnt + 1 > MAX_HOLDERS)
+		return NULL;
+
+	holder = kzalloc(sizeof(TfwMmapBufferHolder) +
+			 sizeof(TfwMMapBufferMem) * num_online_cpus(),
+			 GFP_KERNEL);
+	if (!holder)
+		return NULL;
+
+	order = get_order(size);
+
+	holder->dev_major = -1;
+	holder->size = size;
+	holder->buf = (TfwMmapBuffer **)alloc_percpu_gfp(sizeof(TfwMmapBuffer *),
+							 GFP_KERNEL);
+	if (!holder->buf)
+		goto err;
+
+	atomic_set(&holder->dev_is_opened, 0);
+	atomic_set(&holder->is_freeing, 0);
+
+	page_cnt = size / PAGE_SIZE;
+	/*
+	 * Allocate pages for double mapping and a page for buffer control structure.
+	 */
+	page_ptr = kmalloc(page_cnt * sizeof(struct page *) * 2 + 1, GFP_KERNEL);
+	if (!page_ptr)
+		goto err;
+
+	for_each_online_cpu(cpu) {
+		TfwMmapBuffer *buf, **bufp;
+
+		holder->mem[cpu].buf_page = alloc_pages_node(cpu_to_node(cpu),
+							     GFP_KERNEL, 0);
+		if (holder->mem[cpu].buf_page == NULL)
+			goto err;
+
+		holder->mem[cpu].data_page = alloc_pages_node(cpu_to_node(cpu),
+							      GFP_KERNEL, order);
+		if (holder->mem[cpu].data_page == NULL)
+			goto err;
+
+		page_ptr[0] = holder->mem[cpu].buf_page;
+		for (i = 0; i < page_cnt; ++i) {
+			page_ptr[i + 1] = &holder->mem[cpu].data_page[i];
+			page_ptr[i + page_cnt + 1] = &holder->mem[cpu].data_page[i];
+		}
+
+		buf = (TfwMmapBuffer *)vmap(page_ptr, page_cnt * 2 + 1, VM_MAP, PAGE_KERNEL);
+		if (!buf)
+			goto err;
+
+		buf->size = holder->size;
+		buf->mask = holder->size - 1;
+		buf->head = 0;
+		buf->tail = 0;
+		buf->cpu = cpu;
+		bufp = per_cpu_ptr(holder->buf, cpu);
+		*bufp = buf;
+		atomic_set(&buf->is_ready, 0);
+	}
+
+	if (filename) { /* do not create the file in unit tests */
+		holder->dev_major = register_chrdev(0, filename, &dev_fops);
+		if (holder->dev_major < 0) {
+			T_ERR("Registering char device failed for %s\n", filename);
+			goto err;
+		}
+		holder->dev_class = class_create(THIS_MODULE, filename);
+		if (IS_ERR(holder->dev_class)) {
+			T_ERR("Class creation failed for %s\n", filename);
+			goto err;
+		}
+		holder->dev = device_create(holder->dev_class, NULL,
+					    MKDEV(holder->dev_major, 0),
+					    NULL, filename);
+		if (IS_ERR(holder->dev)) {
+			T_ERR("Device creation failed for %s\n", filename);
+			goto err;
+		}
+		strscpy(holder->dev_name, filename, sizeof(holder->dev_name));
+		holders[holders_cnt++] = holder;
+	}
+
+	kfree(page_ptr);
+
+	return holder;
+
+err:
+	kfree(page_ptr);
+	tfw_mmap_buffer_free(holder);
+
+	return NULL;
+}
+
+void
+tfw_mmap_buffer_free(TfwMmapBufferHolder *holder)
```

Comment:
We don't decrease `holders_cnt` so after creating holders == MAX_HOLDERS we never can create new one

Reply:
Please fix and add test
```
TEST(tfw_mmap_buffer, max_holders_exceeded)
{
	unsigned int i;
	char buf[10];
	
	for (i = 0; i < MAX_HOLDERS; i++) {
		snprintf(buf, sizeof(buf), "test%d", i);
		holders[i] = tfw_mmap_buffer_create(buf, TFW_MMAP_BUFFER_MIN_SIZE);
		EXPECT_NOT_NULL(holders[i]);
	}

	snprintf(buf, sizeof(buf), "test%d", i);
	EXPECT_NULL(tfw_mmap_buffer_create(buf, TFW_MMAP_BUFFER_MIN_SIZE));

	for (i = 0; i < MAX_HOLDERS; i++)
		tfw_mmap_buffer_free(holders[i]);

	holder = tfw_mmap_buffer_create("test1", TFW_MMAP_BUFFER_MIN_SIZE);
	EXPECT_NOT_NULL(holder);
	tfw_mmap_buffer_free(holder);
}
```

Reply:
All buffers are allocated during module initialization and freed upon module unloading. No mechanism for recreating buffers is provided.

Reply:
Currently we destroy buffer on each `tfw_access_log_start`

Reply:
Looks it's redundant. If mmap_buffer is created we need just to return zero.
Fixed.